### PR TITLE
samples: peripheral: 802154_phy_test: Initial Antenna Diversity support

### PR DESCRIPTION
**Pull request by piotrslawecki**
Introduced support for Antenna Diversity feature in IEEE 802.15.4 PHY Test Tool.
Added "lsetcca" command which toggles performing CCA prior to TX.
Added new commands to documentation and fixed minor issues.

Signed-off-by: Piotr Sławęcki <piotr.slawecki@nordicsemi.no>

### CODE REVIEW

**Review by czeslawmakarski — APPROVED** [review]

File: samples/peripheral/802154_phy_test/include/internal/ctrl/ptt_uart.h [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #define UART_CMD_R_SW_VERSION_TEXT "custom rsoftwareversion"
 #define UART_CMD_CHANGE_MODE_TEXT "custom changemode"
 #define UART_CMD_L_GET_CCA_TEXT "custom lgetcca"
+#define UART_CMD_L_SET_CCA_TEXT "custom lsetcca"
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/include/internal/ctrl/ptt_uart.h:39]: It looks you are smuggling one more feature in "Antenna Diversity support" commit. I would split it to a separate commit. If I understand correctly, `lsetcca` is functionally separated from Antenna Diversity.

File: samples/peripheral/802154_phy_test/prj.conf [hunk 2]
```diff
@@ -43,3 +43,7 @@ CONFIG_NCS_SAMPLE_EMPTY_APP_CORE_CHILD_IMAGE=y
 
 # Enable ring buffers
 CONFIG_RING_BUFFER=y
+
+CONFIG_NRF_802154_TEMPERATURE_UPDATE=n
+
+CONFIG_NRF_802154_ENCRYPTION=n
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/prj.conf:49]: I understand that 802.15.4 encryption is disabled, because it is a MAC layer feature that is not needed for PHY testing (it could make some tests hard). But why temperature update is disabled? I think RSSI temperature correction is essential for some of PHY layer features like setting correct Energy Detection threshold, or reporting correct link margin value in broad temperature range.

File: samples/peripheral/802154_phy_test/src/ptt_conf.c [hunk 3]
```diff
@@ -73,11 +74,15 @@ bool ptt_get_ant_mode_ext(uint8_t *ant_mode)
...
-	/* TODO: Implement when antenna diversity is supported in NCS */
+	nrf_802154_antenna_diversity_rx_mode_set(0x02);
+	nrf_802154_antenna_diversity_tx_mode_set(0x02);
 #elif CONFIG_PTT_ANT_MODE_MANUAL
-	/* TODO: Implement when antenna diversity is supported in NCS */
+	nrf_802154_antenna_diversity_rx_mode_set(0x02);
+	nrf_802154_antenna_diversity_tx_mode_set(0x02);
+
+	nrf_802154_antenna_diversity_rx_mode_set(0x01);
+	nrf_802154_antenna_diversity_tx_mode_set(0x01);
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/src/ptt_conf.c:84]: Would it be possible to use some named values instead of magic numbers, as arguments of these functions?

File: samples/peripheral/802154_phy_test/src/rf/ptt_rf.c [hunk 4]
```diff
@@ -102,6 +110,22 @@ static void ptt_rf_set_default_antenna(void)
...
+{
+#if CONFIG_PTT_ANT_MODE_AUTO
+	nrf_802154_antenna_diversity_rx_mode_set(0x02);
+	nrf_802154_antenna_diversity_tx_mode_set(0x02);
+#elif CONFIG_PTT_ANT_MODE_MANUAL
+	nrf_802154_antenna_diversity_rx_mode_set(0x02);
+	nrf_802154_antenna_diversity_tx_mode_set(0x02);
+
+	nrf_802154_antenna_diversity_rx_mode_set(0x01);
+	nrf_802154_antenna_diversity_tx_mode_set(0x01);
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/src/rf/ptt_rf.c:124]: Would be nice to replace magic numbers with some named values.

File: samples/peripheral/802154_phy_test/src/rf_proc.c [hunk 5]
```diff
@@ -96,8 +104,7 @@ static void rf_ed_detected_fn(struct k_work *work)
 }
 
 #if CONFIG_PTT_ANTENNA_DIVERSITY
-#warning "Antenna diversity enabled but not yet supported in NCS"
-/* TODO: Implement when antenna diversity is supported in NCS */
+#pragma message "Antenna diversity is an experimental feature"
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/src/rf_proc.c:107]: It looks like a GCC extension. Is it safe to use it with other toolchains?

File: samples/peripheral/802154_phy_test/src/rf_proc.c [hunk 6]
```diff
@@ -133,7 +139,26 @@ void rf_uninit(void)
 #if CONFIG_PTT_ANTENNA_DIVERSITY
 static void configure_antenna_diversity(void)
 {
-	/* TODO: Implement when antenna diversity is supported in NCS */
+	nrf_ppi_channel_t ppi_channel;
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/src/rf_proc.c:142]: Would be nice to move this code to generic sdk-nrf modules, like initialization of FEMs is handled now in subsys/mpsl/fem. But it's obviously out of scope of this PR.

**Review by hubertmis — COMMENTED** [review]
The proposed changes are acceptable to me, but I would like at least split into a few commits to show in repository history which patch adds antenna diversity, which adds CCA settings, and which covers unrelated documentation fixes.

If it's not feasible to split commits, I'll approve as it is.

File: samples/peripheral/802154_phy_test/prj.conf (hunk 2 above)
Comment (piotrslawecki) [reply samples/peripheral/802154_phy_test/prj.conf:49]: This is a hotfix for a problem, that as I have been informed by @pkwiek is a known issue. Without this, the PHY Test Tool won't work, as it enters and infinite ASSERT FAILURE and restart loop.

Comment (czeslawmakarski) [reply samples/peripheral/802154_phy_test/prj.conf:49]: Let me shine a bit of light on this: initial implementation of PTT did not use the SENSOR subsystem to obtain the data, but rather probed for the temperature explicitly using the `mpsl_...` call. This simplified approach was chosen to faciliate the use of such information on nRF5340-NET platform. The correct re-implementation using SENSOR was planned altogether with porting to nRF5340-APP platform (where some kind of serialization would be needed), but as we know, exporting this sample to the NCS happened earlier.

Comment (hubertmis) [reply samples/peripheral/802154_phy_test/prj.conf:49]: So what's the plan now? To use `mpsl_...` calls in PTT? Is it implemented? Does PTT add correction to RSSI measurements? Or just provide temperature getter command?

File: samples/peripheral/802154_phy_test/Kconfig [hunk 7]
```diff
@@ -51,7 +51,8 @@ config PTT_HW_VERSION
 	range 0 255
 
 config PTT_ANTENNA_DIVERSITY
-	bool "Enable Antenna Diversity (Not Yet Supported)"
+	depends on !SOC_SERIES_NRF53X
+	bool "Enable Antenna Diversity (Experimental Feature)"
 	default n
```
[the remark this answers is not in the document]
Comment (piotrslawecki) [reply samples/peripheral/802154_phy_test/Kconfig:56]: As per my understanding, this way, if someone decides to use menuconfig, the option to enable AD will not appear, making sure that the user will not be surprised by a number of warnings and errors when trying to compile the sample for nRF53 with AD enabled.

[the remark this answers is not in the document]
Comment (hubertmis) [reply samples/peripheral/802154_phy_test/Kconfig:56]: `depends on` causes that the option will appear or not.
`default X` is used only when given option is shown. And by default `bool` types are set to `n`. There is no need to explicitly instruct Kconfig to set `bool` option by default to `n`.

[the remark this answers is not in the document]
Comment (piotrslawecki) [reply samples/peripheral/802154_phy_test/Kconfig:56]: Alright, now I understand. I'll remove the line.

File: samples/peripheral/802154_phy_test/Kconfig [hunk 8]
```diff
@@ -51,7 +51,8 @@ config PTT_HW_VERSION
 	range 0 255
 
 config PTT_ANTENNA_DIVERSITY
-	bool "Enable Antenna Diversity (Not Yet Supported)"
+	depends on !SOC_SERIES_NRF53X
+	bool "Enable Antenna Diversity (Experimental Feature)"
```
Comment (hubertmis) [inline samples/peripheral/802154_phy_test/Kconfig:55]:
```suggestion
	bool "Enable Antenna Diversity [EXPERIMENTAL]"
```

I believe this is the typical way to mark experimental features in Kconfig

File: samples/peripheral/802154_phy_test/src/rf_proc.c (hunk 5 above)
Comment (hubertmis) [reply samples/peripheral/802154_phy_test/src/rf_proc.c:107]: With the note in Kconfig that this feature is experimental, I'm not sure if it's necessary to add compile-time warning. Maybe we can just remove this like?

File: samples/peripheral/802154_phy_test/README.rst [hunk 9]
```diff
@@ -915,6 +1088,11 @@ FEM support
 
 .. include:: /includes/sample_fem_support.txt
 
+.. note::
+   The sample provides support for the *antenna diversity* feature on the nRF52840.
+   It can be enabled and configured using Kconfig configuration options.
```
[the remark this answers is not in the document]
Comment (piotrslawecki) [reply samples/peripheral/802154_phy_test/README.rst:1093]: I've expanded the note with the specific information which option enables the feature. 

As of now, the feature is functional and tested.

**Review by FrancescoSer — APPROVED** [review]

**Comment by FrancescoSer** [comment]
Is this going to be merged before 1.7? If yes, there should be a changelog entry in the related PR

**Comment by czeslawmakarski** [comment]
@FrancescoSer The corresponding entry in the changelog was already introduced.
